core/src/services/koofr/core.rs

```diff
@@ -0,0 +1,455 @@
+
+            let status = resp.status();
+
+            if status != StatusCode::OK {
+                return Err(parse_error(resp).await?);
+            }
+
+            let bs = resp.into_body().bytes().await?;
+            let resp: TokenResponse =
+                serde_json::from_slice(&bs).map_err(new_json_deserialize_error)?;
+            signer.auth_data = AuthData {
+                token: resp.token,
+                mount_id: "".to_string(),
+            };
+
+            let req = Request::get(format!("{}/api/v2/mounts", self.endpoint,));
+
+            let req = req.header(
+                header::AUTHORIZATION,
+                format!("Token token={}", signer.auth_data.token),
+            );
+
+            let req = req
+                .body(AsyncBody::Empty)
+                .map_err(new_request_build_error)?;
+
+            let resp = self.send(req).await?;
+
+            let status = resp.status();
+
+            if status != StatusCode::OK {
+                return Err(parse_error(resp).await?);
+            }
+
+            let bs = resp.into_body().bytes().await?;
+
+            let resp: MountsResponse =
+                serde_json::from_slice(&bs).map_err(new_json_deserialize_error)?;
+
+            for mount in resp.mounts {
+                if mount.is_primary {
+                    signer.auth_data.mount_id = mount.id;
+                    break;
+                }
+            }
+
+            if signer.auth_data.mount_id.is_empty() {
+                return Err(Error::new(ErrorKind::Unexpected, "No primary mount found"));
+            }
+
+            Ok(signer.auth_data.clone())
+        }
+    }
+}
+
+impl KoofrCore {
+    pub async fn ensure_dir_exists(&self, path: &str) -> Result<()> {
+        let path = build_abs_path(&self.root, path);
+
+        let paths = path.split('/').collect::<Vec<&str>>();
+
+        for i in 0..paths.len() - 1 {
+            let path = paths[..i + 1].join("/");
+            debug!("ensure_dir_exists: {}", path);
+            let resp = self.info(&path).await?;
+
+            let status = resp.status();
+
+            match status {
+                StatusCode::NOT_FOUND => {
+                    let resp = self.create_dir(&path).await?;
+
+                    let status = resp.status();
+
+                    match status {
+                        StatusCode::OK | StatusCode::CREATED => {}
+                        _ => return Err(parse_error(resp).await?),
+                    }
+                }
+                StatusCode::OK => {}
+                _ => return Err(parse_error(resp).await?),
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn create_dir(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let paths = path.split('/').collect::<Vec<_>>();
+
+        let name = paths[paths.len() - 1];
+        let relative_path = paths[..paths.len() - 1].join("/");
+        debug!("create_dir: {:?} {} {}", paths, name, relative_path);
```

Comment:
Please don't print debug log in backend.

Reply:
fixed